pnet_sys/src/windows.rs

```diff
@@ -1,112 +1,224 @@
+                    (*storage).sin_family = AF_INET as SockAddrFamily;
+                    (*storage).sin_port = htons(addr.port());
+                    (*storage).sin_addr = inaddr;
+                    mem::size_of::<SockAddrIn>()
+                }
+                SocketAddr::V6(sa) => {
+                    let ip_addr = sa.ip();
+                    let segments = ip_addr.segments();
+                    let inaddr = super::make_in6_addr(segments);
+                    let storage = storage as *mut _ as *mut SockAddrIn6;
+                    (*storage).sin6_family = AF_INET6 as SockAddrFamily6;
+                    (*storage).sin6_port = htons(addr.port());
+                    (*storage).sin6_addr = inaddr;
+                    (*(*storage).u.sin6_scope_id_mut()) = sa.scope_id();
+                    mem::size_of::<SockAddrIn6>()
+                }
+            };
+
+            len as SockLen
+        }
+    }
+
+    pub fn sockaddr_to_addr(storage: &SockAddrStorage, len: usize) -> io::Result<SocketAddr> {
+        unsafe {
+            match storage.ss_family as ctypes::c_int {
+            AF_INET => {
+                assert!(len as usize >= mem::size_of::<SockAddrIn>());
+                let storage: &SockAddrIn = mem::transmute(storage);
+                let ip = super::ipv4_addr(storage.sin_addr);
+                let a = (ip >> 24) as u8;
+                let b = (ip >> 16) as u8;
+                let c = (ip >> 8) as u8;
+                let d = ip as u8;
+                let sockaddrv4 = SocketAddrV4::new(Ipv4Addr::new(a, b, c, d), ntohs(storage.sin_port));
+                Ok(SocketAddr::V4(sockaddrv4))
+            }
+            AF_INET6 => {
+                assert!(len as usize >= mem::size_of::<SockAddrIn6>());
+                let storage: &SockAddrIn6 = mem::transmute(storage);
+                let arr: [u16; 8] = mem::transmute(*storage.sin6_addr.u.Word());
+                let a = ntohs(arr[0]);
+                let b = ntohs(arr[1]);
+                let c = ntohs(arr[2]);
+                let d = ntohs(arr[3]);
+                let e = ntohs(arr[4]);
+                let f = ntohs(arr[5]);
+                let g = ntohs(arr[6]);
+                let h = ntohs(arr[7]);
+                let ip = Ipv6Addr::new(a, b, c, d, e, f, g, h);
+                Ok(SocketAddr::V6(SocketAddrV6::new(ip,
+                                                    ntohs(storage.sin6_port),
+                                                    u32::from_be(storage.sin6_flowinfo),
+                                                    *storage.u.sin6_scope_id())))
+            }
+            _ => Err(io::Error::new(io::ErrorKind::InvalidData, "expected IPv4 or IPv6 socket")),
+        }
+        }
+        
     }
 }
 
 use self::public::*;
 
 #[inline(always)]
 pub fn ipv4_addr(addr: InAddr) -> u32 {
-    (addr.S_un as u32).to_be()
+    unsafe {
+        (*addr.S_un.S_addr() as u32).to_be()
+    }
```

Comment:
This wasn't unsafe before - why has this changed?

Reply:
See comment above. the S_addr() method for the struct field is generated via a new macro in winapi 3, as this is supposed to be a union in future rust versions. So this method is actually unsafe, in future rust versions, I believe this will be unnecessary, I think, due to native union support.